alf/algorithms/muzero_algorithm.py

```diff
@@ -576,26 +656,31 @@ def _split2(x):
 
         return exp1, exp2
 
-    def _reanalyze1(self, replay_buffer: ReplayBuffer, env_ids, positions,
-                    mcts_state_field):
+    def _reanalyze1(self,
+                    replay_buffer: ReplayBuffer,
+                    env_ids,
+                    positions,
+                    mcts_state_field,
+                    horizon: Optional[int] = None):
         """Reanalyze one batch.
 
         This means:
-        1. Re-plan the policy using MCTS for n1 = 1 + num_unroll_steps to get fresh policy
+        1. Re-plan the policy using MCTS for n1 = 1 + horizon to get fresh policy
```

Comment:
Somehow my comment on this function was lost in the last review.
The main issue is that new the logic for handling episode end needs to be changed. For example, suppose mini_batch_length is 5 and step 2 is a LAST step. The target for step 0, 1 and 2 should be calculated using the experience from the first episode. While the target for step 3 and 4 should be based on experience from the next episode. For the current code, the target for all the steps are based on the experience from the first episode.